executor/insert_common.go

```diff
@@ -274,37 +274,53 @@ func (e *InsertValues) fillDefaultValues(row types.DatumRow, hasValue []bool) er
 	return nil
 }
 
-func (e *InsertValues) getRowsSelectChunk(ctx context.Context, cols []*table.Column) ([]types.DatumRow, error) {
+func (e *InsertValues) getRowsSelectChunk(ctx context.Context, cols []*table.Column, exec func(rows []types.DatumRow) error) error {
 	// process `insert|replace into ... select ... from ...`
 	selectExec := e.children[0]
 	if selectExec.Schema().Len() != len(cols) {
-		return nil, ErrWrongValueCountOnRow.GenByArgs(1)
+		return ErrWrongValueCountOnRow.GenByArgs(1)
 	}
-	var rows []types.DatumRow
 	fields := selectExec.retTypes()
+	maxChunkSize := e.ctx.GetSessionVars().MaxChunkSize
 	for {
 		chk := selectExec.newChunk()
 		iter := chunk.NewIterator4Chunk(chk)
 
 		err := selectExec.Next(ctx, chk)
 		if err != nil {
-			return nil, errors.Trace(err)
+			return errors.Trace(err)
 		}
 		if chk.NumRows() == 0 {
 			break
 		}
 
+		numRows := 0
+		if maxChunkSize > chk.NumRows() {
+			numRows = chk.NumRows()
+		} else {
+			numRows = maxChunkSize
+		}
+		rows := make([]types.DatumRow, 0, numRows)
 		for innerChunkRow := iter.Begin(); innerChunkRow != iter.End(); innerChunkRow = iter.Next() {
 			innerRow := innerChunkRow.GetDatumRow(fields)
 			e.rowCount = uint64(len(rows))
```

Comment:
set e.rowCount to `len(rows)` is not correct?

Reply:
The row count here is used to set last insert id of the auto-increment column, which needs a double check for it.